Panic when receiving message on cluster with large number of peers

Found the following panic while trying to reproduce https://github.com/nats-io/gnatsd/issues/954

#### Versions of `gnatsd` and affected client libraries used:

- master (https://github.com/nats-io/gnatsd/commit/a4af582dfb2a9a8a6dc7b0f11da91f25da763410)

#### Steps or code to reproduce the issue:

Trying to create a cluster of a very large size (28 nodes cluster) with a server acting as the seed, then have at least one client connected to each of the servers and send a single message (installed `parallel` on Mac via `brew install parallel`).

```
$ gnatsd -m 8222 -p 4222 --cluster nats://0.0.0.0:6222 --routes nats://127.0.0.1:6222

$ seq 23 50 | parallel -j 27 -u 'gnatsd -p 42{} -m 82{} --cluster nats://0.0.0.0:62{} --routes nats://127.0.0.1:6222'

$ seq 23 50 | parallel -j 27 -u 'nats-sub -s nats://127.0.0.1:42{}' foo
Listening on [foo]
...

# Sending this message causes the panic
$ nats-pub -s nats://127.0.0.1:4222 foo 'Hello world'
```
 
#### Result:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x13bce50]

goroutine 78 [running]:
github.com/nats-io/gnatsd/server.(*client).addSubToRouteTargets(0xc0009c8600, 0xc0036fe360)
        /Users/wallyqs/repos/nats-dev/src/github.com/nats-io/gnatsd/server/client.go:2346 +0x90
github.com/nats-io/gnatsd/server.(*client).processMsgResults(0xc0009c8600, 0xc000144000, 0xc00010eb40, 0xc001e2900c, 0xd, 0x1f4, 0xc001e29004, 0x3, 0x1fc, 0x0, ...)
        /Users/wallyqs/repos/nats-dev/src/github.com/nats-io/gnatsd/server/client.go:2398 +0x15b
github.com/nats-io/gnatsd/server.(*client).processInboundClientMsg(0xc0009c8600, 0xc001e2900c, 0xd, 0x1f4)
        /Users/wallyqs/repos/nats-dev/src/github.com/nats-io/gnatsd/server/client.go:2283 +0x345
github.com/nats-io/gnatsd/server.(*client).processInboundMsg(0xc0009c8600, 0xc001e2900c, 0xd, 0x1f4)
        /Users/wallyqs/repos/nats-dev/src/github.com/nats-io/gnatsd/server/client.go:2192 +0x95
github.com/nats-io/gnatsd/server.(*client).parse(0xc0009c8600, 0xc001e29000, 0x1f, 0x200, 0x1f, 0x0)
        /Users/wallyqs/repos/nats-dev/src/github.com/nats-io/gnatsd/server/parser.go:271 +0x14e1
github.com/nats-io/gnatsd/server.(*client).readLoop(0xc0009c8600)
        /Users/wallyqs/repos/nats-dev/src/github.com/nats-io/gnatsd/server/client.go:740 +0x1d5
created by github.com/nats-io/gnatsd/server.(*Server).startGoRoutine
        /Users/wallyqs/repos/nats-dev/src/github.com/nats-io/gnatsd/server/server.go:1872 +0x91
```

--- a/server/client.go
+++ b/server/client.go
@@ -2352,15 +2352,15 @@ func (c *client) addSubToRouteTargets(sub *subscription) {
 		}
 	}
 
+	var rt *routeTarget
+	lrts := len(c.in.rts)
+
 	// If we are here we do not have the sub yet in our list
 	// If we have to grow do so here.
-	if len(c.in.rts) == cap(c.in.rts) {
+	if lrts == cap(c.in.rts) {
 		c.in.rts = append(c.in.rts, routeTarget{})
 	}
 
-	var rt *routeTarget
-
-	lrts := len(c.in.rts)
 	c.in.rts = c.in.rts[:lrts+1]
 	rt = &c.in.rts[lrts]
 	rt.sub = sub

--- a/server/routes_test.go
+++ b/server/routes_test.go
@@ -20,6 +20,7 @@ import (
 	"strconv"
 	"strings"
 	"sync"
+	"sync/atomic"
 	"testing"
 	"time"
 
@@ -1024,4 +1025,52 @@ func TestRouteSendLocalSubsWithLowMaxPending(t *testing.T) {
 
 	// Check that all subs have been sent ok
 	checkExpectedSubs(t, numSubs, srvA, srvB)
+}
+
+func TestRouteNoCrashOnAddingSubToRoute(t *testing.T) {
+	opts := DefaultOptions()
+	s := RunServer(opts)
+	defer s.Shutdown()
+
+	numRoutes := routeTargetInit + 2
+	total := int32(numRoutes)
+	count := int32(0)
+	ch := make(chan bool, 1)
+	cb := func(_ *nats.Msg) {
+		if n := atomic.AddInt32(&count, 1); n == total {
+			ch <- true
+		}
+	}
+
+	var servers []*Server
+	servers = append(servers, s)
+
+	seedURL := fmt.Sprintf("nats://%s:%d", opts.Cluster.Host, opts.Cluster.Port)
+	for i := 0; i < numRoutes; i++ {
+		ropts := DefaultOptions()
+		ropts.Routes = RoutesFromStr(seedURL)
+		rs := RunServer(ropts)
+		defer rs.Shutdown()
+		servers = append(servers, rs)
+
+		// Create a sub on each routed server
+		nc := natsConnect(t, fmt.Sprintf("nats://%s:%d", ropts.Host, ropts.Port))
+		defer nc.Close()
+		natsSub(t, nc, "foo", cb)
+	}
+	checkClusterFormed(t, servers...)
+
+	// Make sure all subs are registered in s.
+	checkFor(t, time.Second, 15*time.Millisecond, func() error {
+		if s.globalAccount().TotalSubs() != int(numRoutes) {
+			return fmt.Errorf("Not all %v routed subs were registered", numRoutes)
+		}
+		return nil
+	})
+
+	pubNC := natsConnect(t, fmt.Sprintf("nats://%s:%d", opts.Host, opts.Port))
+	defer pubNC.Close()
+	natsPub(t, pubNC, "foo", []byte("hello world!"))
+
+	waitCh(t, ch, "Did not get all messages")
 }

